pipelinedp4j/main/com/google/privacy/differentialprivacy/pipelinedp4j/spark/SparkTable.kt

```diff
@@ -149,7 +167,40 @@ class SparkTable<K, V>(val data: Dataset<Tuple2<K, V>>,
         return filterKeys(stageName) {k -> k in allowedKeysHashSet}
     }
 
+    /**
+     * Randomly samples values per key. The output table will contain same keys as this table, each key will appear
+     * only once. The number of values per key will be at most [count].
+     * It uses window partition by function which requires an extra shuffle and sort operation and introduces
+     * an extra step to transfer data over network but is a scalable and efficient approach for large dataset.
+     */
     override fun samplePerKey(stageName: String, count: Int): SparkTable<K, List<V>> {
-        TODO("Not yet implemented")
+        val listEncoder = Encoders.kryo(List::class.java) as org.apache.spark.sql.Encoder<List<V>>
+        val withRandValueEncoder = Encoders.tuple(keyEncoder, valueEncoder, Encoders.DOUBLE())
+        val withRandomValueEncoder = Encoders.tuple(keyEncoder, valueEncoder, Encoders.INT())
```

Comment:
but it is not used for random value, it is used to encode row number. Should we rename it?